src/browsers/tob/shared/components/DotplotHeatmap.js

```diff
@@ -0,0 +1,241 @@
+import React, { useMemo } from 'react'
+import PropTypes from 'prop-types'
+import { scaleBand, scaleSqrt, scaleSequential, interpolateRgb, extent } from 'd3'
+
+const tileSpacing = 0.05
+
+const DotplotHeatmap = ({
+  id,
+  data,
+  onClick,
+  options: { title, width, height, margin, accessors, xScale, yScale, colorScale, sizeScale },
+}) => {
+  const innerHeight = height - margin.top - margin.bottom
+  const innerWidth = width - margin.left - margin.right
+  const xScaleLocal = useMemo(
+    () =>
+      xScale ||
+      scaleBand()
+        .range([0, innerWidth])
+        .domain([...new Set(data.map((item) => accessors.x(item)))])
+        .padding(tileSpacing),
+    [xScale, data]
+  )
+
+  const yScaleLocal = useMemo(
+    () =>
+      yScale ||
+      scaleBand()
+        .range([innerHeight, 0])
+        .domain([...new Set(data.map((item) => accessors.y(item)))])
+        .padding(tileSpacing),
+    [yScale, data]
+  )
+
+  const colorScaleLocal = useMemo(
+    () =>
+      colorScale ||
+      scaleSequential()
+        .interpolator(interpolateRgb('#FFFFFF', '#CC0000'))
+        .domain(extent(data, accessors.color))
+        .range(['#FFFFFF', '#CC0000'])
+        .nice()
+  )
+
+  const sizeScaleLocal = useMemo(
+    () =>
+      sizeScale ||
+      scaleSqrt()
+        .domain(extent(data, accessors.size).reverse())
+        .range([0, Math.min(xScaleLocal.bandwidth() / 2, yScaleLocal.bandwidth() / 2)])
+        .nice()
+  )
```

Comment:
Add `[sizeScale, data, accessors.size, xScaleLocal, yScaleLocal]` as the dependency list